daft/sql/sql_scan.py

```diff
@@ -203,18 +202,83 @@ def _get_partition_bounds_and_strategy(self, num_scan_tasks: int) -> tuple[list[
         return bounds, strategy
 
     def _single_scan_task(self, pushdowns: Pushdowns, total_rows: int | None, total_size: float) -> Iterator[ScanTask]:
-        file_format_config = FileFormatConfig.from_database_config(DatabaseSourceConfig(self.sql, self.conn))
-        return iter(
-            [
-                ScanTask.sql_scan_task(
-                    url=self.conn.url,
-                    file_format=file_format_config,
-                    schema=self._schema._schema,
-                    num_rows=total_rows,
-                    storage_config=self.storage_config,
-                    size_bytes=math.ceil(total_size),
-                    pushdowns=pushdowns,
-                    stats=None,
-                )
-            ]
+        return iter([self._construct_scan_task(pushdowns, num_rows=total_rows, size_bytes=math.ceil(total_size))])
+
+    def _construct_scan_task(
+        self,
+        pushdowns: Pushdowns,
+        num_rows: int | None = None,
+        size_bytes: int | None = None,
+        partition_bounds: tuple[str, str] | None = None,
+        stats: PyTable | None = None,
+    ) -> ScanTask:
+        if pushdowns.filters is not None:
+            # If the predicate can be translated to SQL, we can apply all pushdowns to the SQL query
+            predicate_sql = pushdowns.filters.to_sql()
+            apply_pushdowns_to_sql = predicate_sql is not None
+        else:
+            # If we don't have a predicate, we can still apply the limit and projection to the SQL query
+            predicate_sql = None
+            apply_pushdowns_to_sql = True
+
+        # If the user has disabled pushdowns, we don't apply any pushdowns to the SQL query
+        apply_pushdowns_to_sql = apply_pushdowns_to_sql and not self._disable_pushdowns_to_sql
+
```

Comment:
Could all this logic be simplified to:

```python
predicate_sql = pushdowns.filters.to_sql() if pushdowns.filters is not None else None
apply_pushdowns_to_sql = not self._disable_pushdowns_to_sql and predicate_sql is not None
```

Reply:
Also another idea: why not just pass `pushdowns` directly into `self._construct_sql_query` instead of needing to do all this logic here?

Nothing about this logic is specific to `_construct_scan_task` really, and if we wanted to share `self._construct_sql_query` with other parts of the codebase (e.g. potentially calling it during the display of a SQLScan logical node) then it would be helpful to be able to have `self._construct_sql_query` also be able to do this logic.

Reply:
1. I'd still want to apply pushdowns to sql if there are no filter pushdowns, i.e. there's only a limit pushdown. So `apply_pushdowns_to_sql = not self._disable_pushdowns_to_sql and predicate_sql is not None` would not be completely correct. Nevertheless I think i can simplify my existing conditions a bit more.
2. I put the `apply_pushdowns_to_sql` logic in this function because it determines whether or not to pass the pushdowns into the scan task https://github.com/Eventual-Inc/Daft/pull/2144/files#diff-af5e4363ffe7b032dfc6346b4fb98a9c0f33dd895f6e65f08b21a22d582dfe72R245 . If we already applied pushdowns to the sql query, then there's no need to pass in the pushdown expressions anymore.

Reply:
Also, I think it does make more sense actually to show the sql queries in the physical plan repr instead of optimized logical plan, because `num_scan_tasks` is also shown in physical plan.